readers/mutation_reader.cc

```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
         return false;
     }
-    if (_prev_kind == mutation_fragment_v2::kind::partition_end) {
-        const bool valid = (kind == mutation_fragment_v2::kind::partition_start);
-        if (valid) {
-            _prev_kind = mutation_fragment_v2::kind::partition_start;
-            _prev_pos = pos;
-        }
-        return valid;
-    }
-    auto cmp = position_in_partition::tri_compare(_schema);
-    auto res = cmp(_prev_pos, pos);
-    bool valid = true;
-    if (_prev_kind == mutation_fragment_v2::kind::range_tombstone_change) {
-        valid = res <= 0;
-    } else {
-        valid = res < 0;
-    }
-    if (valid) {
-        _prev_kind = kind;
-        _prev_pos = pos;
-    }
-    return valid;
-}
-bool mutation_fragment_stream_validator::operator()(mutation_fragment::kind kind, position_in_partition_view pos) {
-    return (*this)(to_mutation_fragment_kind_v2(kind), pos);
-}
 
-bool mutation_fragment_stream_validator::operator()(const mutation_fragment_v2& mf) {
-    const auto valid = (*this)(mf.mutation_fragment_kind(), mf.position());
-    if (valid && mf.is_range_tombstone_change()) {
-        _current_tombstone = mf.as_range_tombstone_change().tombstone();
-    }
-    return valid;
-}
-bool mutation_fragment_stream_validator::operator()(const mutation_fragment& mf) {
-    return (*this)(to_mutation_fragment_kind_v2(mf.mutation_fragment_kind()), mf.position());
-}
+    auto valid = true;
 
-bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind kind) {
-    bool valid = true;
+    // Check fragment kind order
     switch (_prev_kind) {
         case mutation_fragment_v2::kind::partition_start:
-            valid = kind != mutation_fragment_v2::kind::partition_start;
+            valid &= kind != mutation_fragment_v2::kind::partition_start;
             break;
         case mutation_fragment_v2::kind::static_row: // fall-through
         case mutation_fragment_v2::kind::clustering_row: // fall-through
         case mutation_fragment_v2::kind::range_tombstone_change:
-            valid = kind != mutation_fragment_v2::kind::partition_start &&
+            valid &= kind != mutation_fragment_v2::kind::partition_start &&
                     kind != mutation_fragment_v2::kind::static_row;
             break;
         case mutation_fragment_v2::kind::partition_end:
-            valid = kind == mutation_fragment_v2::kind::partition_start;
+            valid &= kind == mutation_fragment_v2::kind::partition_start;
             break;
     }
-    if (valid) {
-        _prev_kind = kind;
+    if (!valid) {
+        return false;
+    }
+
+    if (pos && _prev_kind != mutation_fragment_v2::kind::partition_end) {
+        auto cmp = position_in_partition::tri_compare(_schema);
+        auto res = cmp(_prev_pos, *pos);
+        if (_prev_kind == mutation_fragment_v2::kind::range_tombstone_change) {
+            valid &= res <= 0;
+        } else {
+            valid &= res < 0;
```

Comment:
here too it doesn't seem like we need `valid &=`

Reply:
See above.

Reply:
we return immediately below `if (!valid)`, so there's no need for `&=` here too.

Reply:
@denesb why is `&=` needed here?
`valid` always starts as `true` before entering this block and we always return immediately if it's `false` after the block.